### PR TITLE
OPC-UA Driver

### PR DESCRIPTION
**Pull request by amitjoy**

### CODE REVIEW

File: kura/org.eclipse.kura.driver.opcua.localization/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: OPC-UA Driver Localization Resource
+Bundle-SymbolicName: org.eclipse.kura.driver.opcua.localization;singleton:=true
+Bundle-Version: 1.0.0.qualifier
+Bundle-Vendor: EUROTECH
```
Comment (ctron) [inline kura/org.eclipse.kura.driver.opcua.localization/META-INF/MANIFEST.MF]: This has to be "Eclipse Kura"

File: kura/org.eclipse.kura.driver.opcua.provider/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,34 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: OPC-UA Driver
+Bundle-SymbolicName: org.eclipse.kura.driver.opcua.provider;singleton:=true
+Bundle-Version: 1.0.0.qualifier
+Bundle-Vendor: EUROTECH
```
Comment (ctron) [inline kura/org.eclipse.kura.driver.opcua.provider/META-INF/MANIFEST.MF]: This has to be "Eclipse Kura"

**Review by ctron — CHANGES_REQUESTED** [review]

**Comment by ctron** [comment]
To me it looks like as if the implementation actually polls OPC UA manually. Is this really the case? And, if so, what speaks against using OPC UA event driven model?

**Comment by amitjoy** [comment]
For now, it has been implemented to cope up with Asset request-response paradigm in which every data receive by an Asset in a Wire Graph would trigger a read/write on OPC-UA. The `monitor` functionality of OPC-UA  has not been incorporated and hence left for future enhancement.

**Review by ctron — APPROVED** [review]

**Comment by nicolatimeus** [comment]
I've tested the driver and unfortunately I wasn't able to get it working, this exception gets thrown when the driver tries to read a channel:
```
java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: Could not initialize class org.eclipse.milo.opcua.stack.core.serialization.DataTypeEncoding
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at org.eclipse.kura.internal.driver.opcua.OpcUaDriver.read(OpcUaDriver.java:315)
	at org.eclipse.kura.asset.provider.BaseAsset.read(BaseAsset.java:450)
	at org.eclipse.kura.internal.wire.asset.WireAsset.readChannels(WireAsset.java:303)
	at org.eclipse.kura.internal.wire.asset.WireAsset.onWireReceive(WireAsset.java:232)
	at org.eclipse.kura.internal.wire.helper.WireSupportImpl.updated(WireSupportImpl.java:137)
	at org.eclipse.kura.internal.wire.asset.WireAsset.updated(WireAsset.java:429)
	at org.eclipse.equinox.internal.wireadmin.WireImpl.update(WireImpl.java:190)
	at org.eclipse.kura.internal.wire.helper.WireSupportImpl.emit(WireSupportImpl.java:94)
	at org.eclipse.kura.internal.wire.timer.EmitJob.execute(EmitJob.java:63)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.eclipse.milo.opcua.stack.core.serialization.DataTypeEncoding
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.encodeAsByteString(ExtensionObject.java:99)
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.encode(ExtensionObject.java:95)
	at org.eclipse.milo.opcua.sdk.client.ClientSessionManager.lambda$activateSession$24(ClientSessionManager.java:355)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
	at org.eclipse.milo.opcua.sdk.client.ClientSessionManager.activateSession(ClientSessionManager.java:371)
	at org.eclipse.milo.opcua.sdk.client.ClientSessionManager.lambda$createSession$22(ClientSessionManager.java:320)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

I also found a minor issue when trying to connect to a milo server instance that was listening on the following endpoint:
```
opc.tcp://127.0.0.1:12685
```
In order to connect to the endpoint above the server.name parameter in the driver configuration must be left empty, but this is not currently possible because that parameter is set to required.
I managed to connect to that endpoint by declaring that parameter as optional in the metatype and by removing the trailing '/' that is always appended to the URI by the driver.

**Comment by cdealti** [comment]
Packaging the third party jars (Guava, Milo) is not ideal.

Milo 0.1.0 has been released and available from Maven and we should consume it from there.

No CQ is needed for Milo and it's Guava dependencies.
If Kura code (none as now) needs to to call Guava, then a CQ will be needed.

Please move the dependencies in the target platform and wrap them as plugins where this is needed.

**Comment by amitjoy** [comment]
@cdealti I am okay with providing the required third-party jars as bundles. This OPC-UA library needs to use Milo dependency which internally uses other transitive dependencies which all together result in  **_6.6 MB_** of extra bundles in the runtime and that's why these jars have been kept separated a long time ago.

We also have an imminent problem. This OPC-UA Driver needs the following dependencies:

- Milo Stack Client
- Milo SDK Client
- Milo Stack Core
- Milo SDK Core
- jOOL
- Guava
- Netty ALL [1]

Amongst these libs, `Netty ALL` is not a bundle at all which result in Milo not being resolved at runtime. We could work around these problem by converting `Netty ALL` to a bundle but this would also constrain us from consuming directly from maven.

[1] - https://mvnrepository.com/artifact/io.netty/netty-all

**Comment by amitjoy** [comment]
@nicolatimeus Could you try this now as I have made few changes regarding its class loading issues? Also resolve all the dependencies at your local runtime as it has few more dependencies to few equinox specific bundles.

**Comment by nicolatimeus** [comment]
I've tested the driver again, but still no luck: this exception gets thrown on connection:

```
2017-02-13 15:34:52,977 [DefaultQuartzScheduler_Worker-2] ERROR o.e.k.i.w.a.WireAsset - Error while performing read from the Wire Asset...
org.eclipse.kura.KuraException: "Connection failed. {0}"
	at org.eclipse.kura.asset.provider.BaseAsset.read(BaseAsset.java:452)
	at org.eclipse.kura.internal.wire.asset.WireAsset.readChannels(WireAsset.java:303)
	at org.eclipse.kura.internal.wire.asset.WireAsset.onWireReceive(WireAsset.java:232)
	at org.eclipse.kura.internal.wire.helper.WireSupportImpl.updated(WireSupportImpl.java:137)
	at org.eclipse.kura.internal.wire.asset.WireAsset.updated(WireAsset.java:429)
	at org.eclipse.equinox.internal.wireadmin.WireImpl.update(WireImpl.java:190)
	at org.eclipse.kura.internal.wire.helper.WireSupportImpl.emit(WireSupportImpl.java:94)
	at org.eclipse.kura.internal.wire.timer.EmitJob.execute(EmitJob.java:63)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.eclipse.kura.driver.Driver$ConnectionException: java.util.concurrent.ExecutionException: io.netty.channel.ChannelException: Unable to create Channel from class class io.netty.channel.socket.nio.NioSocketChannel
	at org.eclipse.kura.internal.driver.opcua.OpcUaDriver.connect(OpcUaDriver.java:265)
	at org.eclipse.kura.internal.driver.opcua.OpcUaDriver.read(OpcUaDriver.java:371)
	at org.eclipse.kura.asset.provider.BaseAsset.read(BaseAsset.java:450)
	... 9 more
Caused by: java.util.concurrent.ExecutionException: io.netty.channel.ChannelException: Unable to create Channel from class class io.netty.channel.socket.nio.NioSocketChannel
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at org.eclipse.kura.internal.driver.opcua.OpcUaDriver.connect(OpcUaDriver.java:238)
	... 11 more
Caused by: io.netty.channel.ChannelException: Unable to create Channel from class class io.netty.channel.socket.nio.NioSocketChannel
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:40)
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:321)
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect(Bootstrap.java:163)
	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:145)
	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:126)
	at org.eclipse.milo.opcua.stack.client.UaTcpStackClient.bootstrap(UaTcpStackClient.java:373)
	at org.eclipse.milo.opcua.stack.client.ClientChannelManager.connect(ClientChannelManager.java:190)
	at org.eclipse.milo.opcua.stack.client.ClientChannelManager.getChannel(ClientChannelManager.java:64)
	at org.eclipse.milo.opcua.stack.client.UaTcpStackClient.sendRequest(UaTcpStackClient.java:126)
	at org.eclipse.milo.opcua.stack.client.UaTcpStackClient.getEndpoints(UaTcpStackClient.java:430)
	... 12 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class io.netty.channel.DefaultChannelId
	at io.netty.channel.AbstractChannel.newId(AbstractChannel.java:107)
	at io.netty.channel.AbstractChannel.<init>(AbstractChannel.java:79)
	at io.netty.channel.nio.AbstractNioChannel.<init>(AbstractNioChannel.java:84)
	at io.netty.channel.nio.AbstractNioByteChannel.<init>(AbstractNioByteChannel.java:54)
	at io.netty.channel.socket.nio.NioSocketChannel.<init>(NioSocketChannel.java:98)
	at io.netty.channel.socket.nio.NioSocketChannel.<init>(NioSocketChannel.java:88)
	at io.netty.channel.socket.nio.NioSocketChannel.<init>(NioSocketChannel.java:81)
	at io.netty.channel.socket.nio.NioSocketChannel.<init>(NioSocketChannel.java:74)
	at sun.reflect.GeneratedConstructorAccessor33.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.lang.Class.newInstance(Class.java:442)
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:38)
	... 21 more
```

**Comment by MMaiero** [comment]
@amitjoy It seems that the branch causes some troubles in Travis.
Could you please give it a check?
Furthermore, it is even better if you could squash the commits since most of them provide a single feature. Thanks.

**Comment by amitjoy** [comment]
According to the log, Travis failed while building Karaf Feature for Kura. It noticed a missing package dependency that I am using in this OPC-UA Driver. This package is actually exported by `org.eclipse.osgi` bundle. 

`org.eclipse.osgi.service.urlconversion; version="1.0.0" <org.eclipse.osgi_3.11.1.v20160708-1632 [0]>`

@ctron Could you kindly give us a pointer on this?

**Comment by ctron** [comment]
Karaf may not use Equinox, so you may need to provide an alternative implementation.

**Comment by MMaiero** [comment]
@amitjoy Could you please remove opcua from the distribution? It will be released as dp, so we don't need to have it preinstalled.
Furthermore, could you try to squash the commits into self-contained ones?

**Comment by amitjoy** [comment]
Done 👍 . Relevant commits are squashed as well.

**Comment by MMaiero** [comment]
@amitjoy Thanks! I'll merge it as soon as M1 is released.

**Pull request merged**